rvgo/test/syscall_test.go

```diff
@@ -0,0 +1,903 @@
+
+		require.Equal(t, pc+4, state.PC) // PC must advance
+		require.Equal(t, uint64(0), state.Heap)
+		require.Equal(t, uint64(0), state.LoadReservation)
+		require.Equal(t, uint8(0), state.ExitCode) // ExitCode must be set
+		require.Equal(t, false, state.Exited)      // Must be exited
+		require.Equal(t, preStateRoot, state.Memory.MerkleRoot())
+		require.Equal(t, step+1, state.Step) // Step must advance
+		require.Equal(t, expectedRegisters, state.Registers)
+
+		fastPost := state.EncodeWitness()
+		runEVM(t, contracts, addrs, stepWitness, fastPost)
+		runSlow(t, stepWitness, fastPost, nil)
+	})
+}
+
+func FuzzStateSyscallGetrlimit(f *testing.F) {
+	contracts := testContracts(f)
+	addrs := testAddrs
+
+	testGetrlimit := func(t *testing.T, addr, pc, step uint64) {
+		pc = pc & 0xFF_FF_FF_FF_FF_FF_FF_FC // align PC
+		addr = addr &^ 31
+		state := &fast.VMState{
+			PC:              pc,
+			Heap:            0,
+			ExitCode:        0,
+			Exited:          false,
+			Memory:          fast.NewMemory(),
+			LoadReservation: 0,
+			Registers:       [32]uint64{17: fast.SysGetrlimit, 10: 7, 11: addr},
+			Step:            step,
+		}
+		state.Memory.SetUnaligned(pc, syscallInsn)
+		expectedRegisters := state.Registers
+		expectedRegisters[10] = 0
+		expectedRegisters[11] = 0
+
+		fastState := fast.NewInstrumentedState(state, nil, os.Stdout, os.Stderr)
+		stepWitness, err := fastState.Step(true)
+		require.NoError(t, err)
+		require.False(t, stepWitness.HasPreimage())
+
+		postMemory := fast.NewMemory()
+		postMemory.SetUnaligned(pc, syscallInsn)
+		var bytes [8]byte
+		binary.LittleEndian.PutUint64(bytes[:], 1024)
+		postMemory.SetUnaligned(addr, bytes[:])
+		postMemory.SetUnaligned(addr+8, bytes[:])
+
+		require.Equal(t, pc+4, state.PC) // PC must advance
+		require.Equal(t, uint64(0), state.Heap)
+		require.Equal(t, uint64(0), state.LoadReservation)
+		require.Equal(t, uint8(0), state.ExitCode) // ExitCode must be set
+		require.Equal(t, false, state.Exited)      // Must be exited
+		require.Equal(t, state.Memory.MerkleRoot(), postMemory.MerkleRoot())
+		require.Equal(t, step+1, state.Step) // Step must advance
+		require.Equal(t, expectedRegisters, state.Registers)
+
+		fastPost := state.EncodeWitness()
+		runEVM(t, contracts, addrs, stepWitness, fastPost)
+		runSlow(t, stepWitness, fastPost, nil)
+	}
+
+	testGetrlimitErr := func(t *testing.T, res, addr, pc, step uint64) {
+		pc = pc & 0xFF_FF_FF_FF_FF_FF_FF_FC // align PC
+		addr = addr &^ 31
+		state := &fast.VMState{
+			PC:              pc,
+			Heap:            0,
+			ExitCode:        0,
+			Exited:          false,
+			Memory:          fast.NewMemory(),
+			LoadReservation: 0,
+			Registers:       [32]uint64{17: fast.SysGetrlimit, 10: res, 11: addr},
+			Step:            0,
+		}
+		state.Memory.SetUnaligned(pc, syscallInsn)
+
+		fastState := fast.NewInstrumentedState(state, nil, os.Stdout, os.Stderr)
+		_, err := fastState.Step(true)
+		require.Contains(t, err.Error(), "f0012")
```

Comment:
We should remove this hardcoded errorcode in the future.

Reply:
yeah we should. let me take it to the next task.